### DISCUSSION TITLE
RFC: Passing styles / theme to blocks

### DISCUSSION THREAD

**Discussion by CiaranMn** [blockprotocol/blockprotocol#180] [RFC]
## Goals
- Allow blocks to apply styling to make them appear (more) visually consistent with the application embedding them.

## Constraints
- Should **not** force styling onto blocks which might break their visual appearance in some way.
- Should **not** require the application to have any special knowledge of a particular block.
- Should work if the block has its own shadow DOM.

### Current
The spec previously said that:

> Embedding applications SHOULD provide a `styleVariables` object passed to the block alongside other fields, containing any of the theme variables listed in Appendix A which they have a preferred value for.
>
> Blocks SHOULD use these variables where appropriate to make their block visually consistent with the embedding application (except where applying application-supplied values may interfere with block functionality or accessibility).

_* Edited to remove links to defunct spec pages – see discussion below for suggested variables_

This allows embedding applications to pass a set of variables which the block can apply or not apply at will. Blocks can also provide fallback values if the variables are not defined.

**Issues**
- the existing list almost certainly needs expanding with more variables to be effective
  - these could potentially be namespaced, e.g. `{ input: { borderRadius: 8 } }`
- it is cumbersome for blocks to pull out variables and apply them in many places
  - if the provided variables can be mapped to CSS selectors, blocks could automatically generate stylesheets from objects (with any desired fallbacks / omissions)

### Potential expansion / alternative

The embedding application could instead (or additionally) pass an absolute URL stylesheet to a block, which the block then loads. This would add styling data much simpler.

If passed in addition to a style object, blocks could choose whether to load the stylesheet and provide very minimal styling, or to take values from the object instead.

**Issues**
- blocks would have less control over the applied styling and there is a greater potential for clashes between the block's styling objectives and the stylesheet
- the specificity of selectors in the provided stylesheet vs the block's own would make which values are applied more unpredictable
  -  blocks could apply inline styles to elements where important

## Questions
1. Do you think that some kind of object containing theme variables should be passed to blocks?
  a. if so, what should the variables be?
  b. if not, why not? 

2. Should a style sheet be passed to blocks? In addition or instead of an object? Why / why not?

**Comment by MaggieAppleton** [comment]
I'm leaning towards giving the embedding application stylistic authority over how the block presents in their application, rather than the block author. As a block author it will be tricky to predict all the contexts where your block might be used. So even if it looks fabulous in an isolated context, their styling choices aren't going to fit well in all embedding applications.

Which would mean more granular control over CSS properties via a stylesheet. In terms of selector specificity, I'd expect the embedding application's stylesheet to override all declarations in the block. The same way a user stylesheet has ultimate authority in a browser.

As part of the styling guidelines, we should obviously tell block creators to create sensible, logical classnames on major elements of their block to make it easy for embedders to access the CSS properties.

If there a downside to the block author giving up stylistic authority I'm not seeing? Some block creators might be frustrated if people make their blocks look terrible, but it's not harming anything other than their ego (and possible the embedding app's users if they make it an inaccessible mess).

That said, I like the simplicity of passing a pre-defined set of variables, rather than making block embedders hunt down the specific selectors of every element in a block. Also helps make the system more accessible for people with limited CSS skills. So the object + stylesheet approach feels like a good balance.
Object as default, stylesheet option available if you need it.

In terms of variables, here's the current list:
- **primaryColor**: a primary theme color for main action elements and fills
- **secondaryColor:** a secondary color for highlights and variations
- **borderColor**: a color to use for borders for tables, boxes, etc
- **borderRadius**: the radius of borders for tables, boxes, buttons, etc
- **backgroundColor**: a color to use as the background for elements
- **fontColor**: the default text color
- **fontFamily**: the default font

I'd suggest we add on the following properties:
- **tertiaryColor** – gives block creators more flexibility. Also thinking ahead to when people create data visualisation blocks where multiple colours are needed for mapping semantic meaning
- **secondaryBackgroundColor** – (or could be called **neutralColor**) for supporting grayscale elements that are common to create shallow depth in interfaces and signifying active/disabled states. The specific grey value used across a site usually has a significant effect on how it looks and feels.
Eg. see the way gray is used in this Github text box 
![Screenshot - 2022-02-03 09 32 19](https://user-images.githubusercontent.com/5599295/152316478-43bf5f91-df73-4912-8146-4388a0aa4c6d.png)

- **baseFontSize** – starting base font size (1rem = 16px)
- **typeScale** – ratio for scaling up the base font size (eg. 1.125, 1.25, 1.33) - https://type-scale.com/
- **baseSpaceSize** – 8px (8px is standard across most products, but people might want to adjust it)
- **spacingScale** – adjusts how the base spacing size scales up. probably best to do this as an array of values: [0.5, 1, 1.5, 2, 3, 4, 5]
- **borderWeight** – 1, 2, 3
- **iconSize** – applies to all icons in the block. different icon sizes in an app vs. a block would be visually noticeable. This can be a pixel value. Eg. 24px is standard in material design. Most apps use around 16px, but the range varies from 12-32px

I think we should divide border from button radius:
- **boxBorderRadius** – for border radius on box objects (forms, containers, etc.)
- **buttonRadius** – for border radius on buttons specifically. allows people to create rounded buttons without also having all their other elements turn to circles.

I'd be in favour of us including some design guidelines in the BP docs. Nothing comprehensive, just encouraging best practices drawn from established systems like Material Design (readable font sizes, AA+ contrast ratios, sensible and consistent spacing systems, minimum touch target sizes on clickable elements). And emphasising keeping block design fairly neutral to reduce the need for embedding apps to dig into CSS adjustments. If most of our block creators are going to be developers who don't have access to design support, providing a basic set of guidelines could go a long way to improving the quality and usability of blocks.

**Comment by CiaranMn** [comment]
> I'm leaning towards giving the embedding application stylistic authority over how the block presents in their application, rather than the block author. As a block author it will be tricky to predict all the contexts where your block might be used. So even if it looks fabulous in an isolated context, their styling choices aren't going to fit well in all embedding applications.

I agree with this as a principle. My only worry is having blocks visually broken in some way (rather than just looking bad), e.g. if styling applied by the app somehow makes text illegible, UI controls unusable, etc.

I think having blocks keep their design very neutral would help with this.

> In terms of selector specificity, I'd expect the embedding application's stylesheet to override all declarations in the block. The same way a user stylesheet has ultimate authority in a browser.

The easiest way of doing this is for the application's css to have `!important` thrown on everything and require that the block does not.

> As part of the styling guidelines, we should obviously tell block creators to create sensible, logical classnames on major elements of their block to make it easy for embedders to access the CSS properties.

I was thinking more that stylesheet selectors would be at the level of `input`, `button`, rather than classnames - we should figure out what issues might remain after doing that and then consider specifying classnames. Embedding applications shouldn't have to have any knowledge of the internal details of a block, so we'd have to pick a set of defined classnames to apply to things, in the same way as listing style variables. I wonder if a stylesheet _and_ variables to apply _and_ classnames to apply is too much - classnames would just be a combination of values that could be applied via variables instead (and vice versa) so maybe we should pick one of the latter two.

> I'd be in favour of us including some design guidelines in the BP docs. Nothing comprehensive, just encouraging best practices drawn from established systems like Material Design (readable font sizes, AA+ contrast ratios, sensible and consistent spacing systems, minimum touch target sizes on clickable elements).

This would be helpful.

**Comment by vilkinsons** [comment]
I think there's probably quite a bit we could learn from [FAST](https://www.fast.design/docs/design-systems/fast-frame/#configuring-styles) around this ([context](https://www.fast.design/docs/design-systems/design-tokens#fast-frame-design-tokens)).

**Reply by julrich** [reply]
Just skimmed through this discussion. Super interesting! Just wanted to echo that Design Tokens could be a really great baseline for exactly this (re: FAST). They (w3c community group) just released their Second Editors' Draft:
https://www.w3.org/community/design-tokens/2022/06/14/call-to-implement-the-second-editors-draft-and-share-feedback/

It's the approach we're using in our Design System projects, it pretty much single-handedly enables multi-layered Design System approaches... where ~90% of design application happens by those tokens, and only ~10% is stuff overwritten directly (CSS cascade or !important). It successfully decouples design and branding from the component API for us.

The values @MaggieAppleton cites are super relevant, but remind me of the pre-cursors to whats quickly becoming a standard: semantic tokens (https://medium.com/eightshapes-llc/naming-tokens-in-design-systems-9e86c7444676), and a good shared taxonomy.

This can also be translated to a slew of different other formats, for interoperability, through the likes of Style Dictionary, and other related tooling: https://amzn.github.io/style-dictionary/. For example: we built a Bootstrap theme converter for a client, which takes his semantic tokens, and automatically translates theme into a drop-in usable Bootstrap theme as part of his CI/CD.

**Comment by davidebriani** [comment]
I agree with the principle that the application should not have any special knowledge of a particular block. This is important if you have a marketplace scenario for blocks and the embedding application allows users to choose and load blocks from the marketplace dynamically.

Forcing styling onto blocks has downsides since you don't know in advance which styles to apply or which specificity is required; I fear that juggling with those will only lead to more complexity and less maintainability for the app's styles.
The block's author is ultimately responsible for producing a block that is good-looking and theme-able.

IMO, the embedding app could provide a theme containing a set of design tokens defined in the RFC; the theme would be only used as a default reference by the block.

The approach of the CSS URL is not great since:
- it would have to know the block's specifics to apply dedicated styling
- the author of the block still has the faculty of discarding the URL and rendering a different styling in its shadow-dom
- the block cannot easily extract values defined in the CSS file for dynamic processing of other uses

Providing an object, as in the current spec, seems a more flexible option:
- no limits on the block part on how the values can be parsed and used
- more ways in which the embedding app can define and dynamically change the theme that it is providing to the block. Useful when the embedding app is a white-label product and the theme can be defined dynamically but you don't want to build and store a CSS file each time a theme is updated for the app.

The RFC could:
1. Adopt a specification for the structure of the theme object. Here the work of others comes in handy: [Styled System](https://styled-system.com/theme-specification), [SystemUI](https://system-ui.com/theme/).
2. Suggest a precise theme object: i.e. that prescribes the name of the colors to be defined (primary, secondary, etc.), how many different sizes to choose from for the font, how many breakpoints are there for media queries, etc.

There might be cases where the design tokens of the embedding app do not match the ones prescribed by the RFC as in 2. However, the app can still choose to leverage Blockprotocol as much as possible by:
- supporting 1. and translating its design tokens to match 2., thus being able to properly use existing blocks.
- supporting just 1. as a defined and useful way to provide styling, while managing its own marketplace of blocks that expect the same design tokens

**Reply by vilkinsons** [reply]
Thanks, Davide. Agree re: block developers holding ultimate responsibility for producing blocks that are good-looking and theme-able. Maggie Appleton on our team has written up some [design guidelines for block developers](https://hash.dev/blog/block-design) over on the hash.dev blog that aims to provide a starting point for this. I'll let the BP core team get back to you on your other points!

**Comment by jtewright** [comment]
Hi @davidebriani,

I'm the product manager working on the Block Protocol. Thanks for your thoughts and suggestions on this!

I agree with you on the limitations of the CSS URL approach. However, we've also been thinking that it could be easier for block developers because the CSS URL can provide styles for standard selectors (e.g. button, input) which would be applied automatically, whereas developers would have to implement variables from the theme object directly. A solution might be to require certain base styles for standard selectors in the theme object and then generate a CSS URL automatically so block developers can rely on that as a default. Additionally, while developers can discard the CSS URL, they can also discard the theme object. We expect embedding applications to review and filter out low-quality blocks, and we have ideas for things like ratings to help with this.

With regards to your recommendations, that's very aligned with what we've been thinking. We have some work in progress on an expanded theme object which defines more specific variables than we've listed in the [current spec](https://blockprotocol.org/spec/appendix-a-styling). If embedding applications provide a theme object, block developers will be required to use its variables to style their block's UI. We also have ideas for UI kits and other resources to make this as easy as possible.

We still have some questions around this approach. It would be interesting to get your thoughts on these!

1. Which variables or design tokens should we have? How do we balance the goal of covering everything with the goal of keeping things simple? (Thank you for the links you've shared on this.)
2. How easy, or even possible, will it be for developers to correctly implement these variables? E.g. will different block developers agree on which UI elements should have a secondary vs. a tertiary color?
3. Should we provide a way for embedding applications to provide custom styling patches for specific blocks? E.g. for a scenario where a block forgets to implement one specific variable which is particularly important to an embedding application (as far as can be achieved given technical limitations if a block encapsulates its styling in a shadow DOM).

Let us know if you have any thoughts on these questions or other ideas.

Joseph

**Comment by davidebriani** [comment]
Hi there, thanks for the kind reply ;)

The approach with the CSS URL might have the benefit of being simple enough for block developers to get started quickly, and I agree that's a great point.
However, there are some more shortcomings I see:

1. The app must compile or store a CSS file. Sometimes this is not feasible.
2. The styling you can cover with basic CSS is too little. The app wouldn't know how to style complex (a.k.a. real-world) UI components, that is the responsibility of the block's developer, which in turn doesn't have proper tooling to style his components and be sure they don't break when another CSS theme is used.
3. You have to instruct developers on how to correctly inject the CSS file, assuming a shadow-dom is in play. Basically, you're telling them "here is a black-box object, have faith and place it like this in your block. Now you can try to write some other custom CSS to overwrite the parts that don't work well while building your components". Some developers might just shoot `!important` everywhere.
4. Passing CSS around forces the assumption of working in a web / CSSOM context. I reckon that BlockProtocol could easily extend to other platforms such as [React Native](https://reactnative.dev/) / [React Native Web](https://necolas.github.io/react-native-web/).
5. It shouldn't be too hard to pass the object of theme variables and have an NPM package that automatically converts those variables into a basic CSS theme that you can inject if you still want to grant that kind of Developer Experience. As a block developer, I'd prefer this explicit and pick-what-you-need approach over a black-box / implicit one.

Regarding your questions:

> Which variables or design tokens should we have?

It's up to the embedding app to define which design tokens exist. You cannot know in advance how many breakpoints the themes will have, or if they use `accent`, `brand`, `secondary`, `xyz` colors, or the different kinds of spacing that exist.
Blockprotocol can help by prescribing *how* design tokens can be defined and used.

> How do we balance the goal of covering everything with the goal of keeping things simple?

Great tooling. That's also what happened around System UI / Styled System:
- there's the specification that prescribes how to define design tokens
- there's the tool that accepts a theme definition with that format and provides you with a way to easily use it to style components
- there are now many libraries that use that tooling, and in turn accept that format of theme definition. As examples: [Chakra UI](https://chakra-ui.com/docs/styled-system/theming/theme), [Github's Primer](https://primer.style/react/theming#system-props-and-the-sx-prop), [Backlight](https://backlight.dev/docs/design-tokens-using-system-ui).

> How easy will it be for developers to correctly implement these variables? E.g. will different block developers agree on which UI elements should have a secondary vs. a tertiary color?

With great tooling, it shouldn't be too hard.
Since Blockprotocol also hosts a Hub to upload public blocks, to ease the work of *those* developers it could also provide a reference theme  definition with specific tokens they can rely on. Should you choose to use an approach similar to Styled System, chances are that there's little work to be done to convert the theme object into one that is accepted by the UI library that the developer chose to build its block.

> Should we provide a way for embedding applications to provide custom styling patches for specific blocks?

I agree there are cases where an escape hatch would be handy. I don't have great ideas on this yet, since even in a CSSOM context the developer can just hide its styling into a Shadow DOM.

**Reply by jtewright** [reply]
Thanks for the detailed response @davidebriani ! 

Earlier this week we [announced an update](https://hash.dev/blog/block-protocol-v02) to the Block Protocol. It mainly rearchitects the protocol into Core and Service Specifications. We discussed styling a lot as part of the project—the idea is that styling, and other things like it which not all applications/blocks might need, will be specific services. We only have the Graph Service (for data storage) ready at this time, but Styling is listed as a new service we are working on here: https://blockprotocol.org/docs/spec/rfcs_and_roadmap#styling

Might you be interested in collaborating on building this service? If not, perhaps you'd have time to jump on a call to discuss the points you raise when we get to that phase of the project?

Thank you,
Joseph

**Comment by sudheer-gowrigari** [comment]
@jtewright @CiaranMn ,
Hi , The link to [current spec](https://blockprotocol.org/spec/appendix-a-styling). is broken.!! Can you please provide the right link ?
Thanks

**Reply by vilkinsons** [reply]
Hi Sudheer, we've moved most of the conversation around this to the current GitHub discussion thread. The remaining in-spec reference to styling is now under our RFCs and roadmap section, simply pointing users here: https://blockprotocol.org/docs/spec/rfcs_and_roadmap#styling (replacing our old guidance)

Separately, we've published design guidelines for developing blocks here: https://hash.dev/blog/block-design The emphasis is on minimal, neutral appearance (with a view to introducing design tokens in the future that embedding applications may use to customize block's appearances to match their own).

**Comment by davidebriani** [comment]
I stumbled upon the Block Protocol again and, oh wow, how much progress you've been making :)

Regarding themes and styles, I am curious if there are some related developments.

In addition to the conversation above, I wanted to point out the [Design Tokens Format Module](https://tr.designtokens.org/format/) specification, which distills years of work and [different experiments](https://amzn.github.io/style-dictionary) on how to interoperate design tokens between different tools. 

The specification is already being used as a reference for exchanging design tokens on a number of tools and platforms:
- [Figma](https://twitter.com/pwnies/status/1671584984057151489), [Tokens Studio](https://twitter.com/EstherCheran/status/1667440676336807936)
- [Cobalt](https://cobalt-ui.pages.dev/)
- [Universal Design Tokens](https://udt.design/)

So, although this doesn't answer all the questions, it's a base reference that could be considered for the scope of providing styling to blocks.

**Reply by vilkinsons** [reply]
Hi Davide, thanks for circling back on this.

A log of updates can be found over on the [roadmap page](https://blockprotocol.org/roadmap). Much of our core work on the Block Protocol has been on the backburner over the last 12 months as we focus primarily on [HASH](https://hash.ai/).

We're with you on the approach outlined in your post, and agree a version of this looks like the ideal solution.
